File: tests/cpp/test_audio.cpp
#include <gtest/gtest.h>
#include "audio/audio.hpp"

#include <cstdint>
#include <cstdio>
#include <fstream>
#include <string>
#include <vector>

using namespace loom;

// Audio tests run against a dummy audio driver (SDL_AUDIODRIVER=dummy).
// We verify lifecycle and state transitions; no actual sound is produced.

namespace {

// Write a tiny mono 16-bit PCM WAV. The tests below need a file that ACTUALLY
// loads: playing a nonexistent file bails out early and never exercises the
// interesting path (a live sound being started, then released).
std::string write_temp_wav(const char* name) {
    const uint32_t sample_rate = 8000;
    const uint16_t channels    = 1;
    const uint16_t bits        = 16;

    std::vector<int16_t> samples(2000);
    for (size_t i = 0; i < samples.size(); ++i)
        samples[i] = static_cast<int16_t>((i % 20 < 10) ? 3000 : -3000);

    const uint32_t data_bytes  = static_cast<uint32_t>(samples.size() * sizeof(int16_t));
    const uint32_t byte_rate   = sample_rate * channels * bits / 8;
    const uint16_t block_align = channels * bits / 8;

    std::string path(name);
    std::ofstream f(path, std::ios::binary);
    auto u32 = [&](uint32_t v) { f.write(reinterpret_cast<const char*>(&v), 4); };
    auto u16 = [&](uint16_t v) { f.write(reinterpret_cast<const char*>(&v), 2); };

    f.write("RIFF", 4); u32(36 + data_bytes); f.write("WAVE", 4);
    f.write("fmt ", 4); u32(16); u16(1); u16(channels); u32(sample_rate);
    u32(byte_rate); u16(block_align); u16(bits);
    f.write("data", 4); u32(data_bytes);
    f.write(reinterpret_cast<const char*>(samples.data()), data_bytes);
    return path;
}

} // namespace

TEST(Audio, EngineInitializes) {
    // May return initialized=false in headless CI — that's OK, it's non-fatal
    AudioEngine engine;
    // Either it worked or it gracefully degraded
    SUCCEED();
}

TEST(Audio, MusicNotPlayingByDefault) {
    AudioEngine engine;
    EXPECT_FALSE(engine.music_playing());
}

TEST(Audio, StopMusicWhenNotPlaying) {
    AudioEngine engine;
    EXPECT_NO_THROW(engine.stop_music());
}

TEST(Audio, SetMasterVolume) {
    AudioEngine engine;
    EXPECT_NO_THROW(engine.set_master_volume(0.5f));
    EXPECT_NO_THROW(engine.set_master_volume(0.f));
    EXPECT_NO_THROW(engine.set_master_volume(1.f));
}

TEST(Audio, PlayNonExistentFileReturnsEmptyHandle) {
    AudioEngine engine;
    if (!engine.initialized()) GTEST_SKIP() << "Audio not available in this env";

    SoundHandle h = engine.play_sound("nonexistent_file.wav");
    // Should not crash; just return an empty handle
    EXPECT_FALSE(h.playing());
}

TEST(Audio, PlayMusicNonExistentFileNoThrow) {
    AudioEngine engine;
    if (!engine.initialized()) GTEST_SKIP() << "Audio not available in this env";

    EXPECT_NO_THROW(engine.play_music("nonexistent.ogg"));
    EXPECT_FALSE(engine.music_playing());
}

// Regression: play_sound() used to return a shared_ptr<ma_sound> with the
// DEFAULT deleter. Dropping the handle — the normal way to fire a one-shot —
// raw-deleted the ma_sound without calling ma_sound_uninit(), leaving a freed
// but still-registered sound in the engine's node graph. miniaudio's audio
// thread then mixed from released memory and the process crashed.
TEST(Audio, PlaySoundFireAndForgetIsSafe) {
    AudioEngine engine;
    if (!engine.initialized()) GTEST_SKIP() << "Audio not available in this env";

    const std::string path = write_temp_wav("loom2d_test_tone_ff.wav");
    for (int i = 0; i < 32; ++i)
        engine.play_sound(path, 0.05f);   // handle dropped immediately
    SUCCEED();                             // must reach here without crashing
    std::remove(path.c_str());
}

TEST(Audio, PlaySoundReturnsALiveHandle) {
    AudioEngine engine;
    if (!engine.initialized()) GTEST_SKIP() << "Audio not available in this env";

    const std::string path = write_temp_wav("loom2d_test_tone_h.wav");
    SoundHandle h = engine.play_sound(path, 0.05f);
    EXPECT_NE(h.impl, nullptr);
    EXPECT_NO_THROW(h.set_volume(0.5f));
    EXPECT_NO_THROW(h.stop());
    std::remove(path.c_str());
}

// A one-shot must never outlive its engine: the AudioEngine destructor has to
// uninit any still-playing sounds BEFORE ma_engine_uninit().
TEST(Audio, EngineDestroyedWhileSoundStillPlaying) {
    const std::string path = write_temp_wav("loom2d_test_tone_d.wav");
    {
        AudioEngine engine;
        if (!engine.initialized()) {
            std::remove(path.c_str());
            GTEST_SKIP() << "Audio not available in this env";
        }
        engine.play_sound(path, 0.05f);   // still playing as the engine dies
    }
    SUCCEED();
    std::remove(path.c_str());
}

TEST(Audio, MultipleEnginesIndependent) {
    // Two engines can coexist without interfering
    AudioEngine e1, e2;
    EXPECT_NO_THROW({
        e1.set_master_volume(0.8f);
        e2.set_master_volume(0.5f);
    });
}

File: tests/python/test_audio.py
"""Audio bindings: AudioEngine + SoundHandle.

These cover the one-shot SFX path — `audio.play_sound(path)` — which is the most
common audio call in a game and which used to be completely broken from Python:

  * `SoundHandle` was never registered with pybind11, so play_sound() raised on
    return-value conversion, and
  * the ma_sound was held by a shared_ptr with the DEFAULT deleter, so releasing
    it freed the sound without ma_sound_uninit() while miniaudio's audio thread
    was still mixing it — a hard crash.

The tests skip when there's no audio device (headless CI).
"""
import struct
import wave

import pytest

import loom2d as loom


@pytest.fixture
def tone(tmp_path):
    """A short mono 16-bit WAV that actually decodes.

    Playing a *nonexistent* file bails out early and never exercises the
    interesting path, which is exactly how the crash above went unnoticed.
    """
    path = tmp_path / "tone.wav"
    with wave.open(str(path), "w") as w:
        w.setnchannels(1)
        w.setsampwidth(2)
        w.setframerate(8000)
        frames = b"".join(
            struct.pack("<h", 3000 if (i // 10) % 2 else -3000) for i in range(2000)
        )
        w.writeframes(frames)
    return str(path)


@pytest.fixture
def engine():
    e = loom.AudioEngine()
    if not e.initialized:
        pytest.skip("no audio device available in this environment")
    return e


class TestAudioEngine:
    def test_constructs(self):
        loom.AudioEngine()  # never throws, even with no device

    def test_music_not_playing_by_default(self):
        assert loom.AudioEngine().music_playing() is False

    def test_master_volume(self):
        e = loom.AudioEngine()
        e.set_master_volume(0.5)
        e.set_master_volume(0.0)
        e.set_master_volume(1.0)


class TestPlaySound:
    def test_returns_a_sound_handle(self, engine, tone):
        handle = engine.play_sound(tone, 0.05)
        assert isinstance(handle, loom.SoundHandle)

    def test_handle_exposes_controls(self, engine, tone):
        handle = engine.play_sound(tone, 0.05)
        assert isinstance(handle.playing, bool)
        handle.set_volume(0.5)
        handle.stop()

    def test_fire_and_forget_is_safe(self, engine, tone):
        # The handle is optional: the engine keeps the sound alive until it
        # finishes. Dropping it must not free a sound that is still playing.
        for _ in range(32):
            engine.play_sound(tone, 0.05)

    def test_missing_file_returns_a_silent_handle(self, engine):
        handle = engine.play_sound("no_such_file.wav")
        assert handle.playing is False

    def test_handle_may_outlive_the_engine(self, tone):
        # Regression: the sound's deleter runs ma_sound_uninit(), which must not
        # execute against a destroyed ma_engine. Dropping the engine while a
        # handle is still held used to crash, so every sound keeps its engine
        # alive until the sound itself is released.
        engine = loom.AudioEngine()
        if not engine.initialized:
            pytest.skip("no audio device available in this environment")
        handle = engine.play_sound(tone, 0.05)
        del engine          # engine goes first...
        del handle          # ...sound released afterwards: must not crash

    def test_stopped_sounds_do_not_accumulate(self, engine, tone):
        # Sounds stopped early never reach at_end(); reaping on that alone would
        # hold them for the engine's whole lifetime.
        for _ in range(16):
            engine.play_sound(tone, 0.05).stop()


class TestMusic:
    def test_play_and_stop(self, engine, tone):
        engine.play_music(tone, 0.5, False)
        engine.set_music_volume(0.2)
        engine.stop_music()
        assert engine.music_playing() is False

    def test_missing_file_does_not_play(self, engine):
        engine.play_music("no_such_file.wav")
        assert engine.music_playing() is False

    def test_stop_when_not_playing(self, engine):
        engine.stop_music()
